### ISSUE TITLE
Add stat scaling info to rendered output

### ISSUE THREAD

**Issue by pastelmind** [joeistas/heroes-parser#5]
Currently (v1.0.11),  heroes-parser does not include stat scaling information to rendered tooltips. Stat scaling is an important piece of information, and adding the info would be beneficial to end users.

Current output:

```html
After a <span>0.5</span> second delay, enemies in front of Alarak take <span>175</span> damage and are silenced for <span>1.5</span> seconds. 
```

Desired output:

```html
After a <span>0.5</span> second delay, enemies in front of Alarak take <span>175 (+4%)</span> damage and are silenced for <span>1.5</span> seconds. 
```

To do this, I expect `renderTooltipData()` would need to test each formula when parsing references. If a formula consists of only 1 value (no arithmetic operators), the function should search for a `<LevelScalingArray>` whose `<Entry>` and `<Field>` matches the value. If such an element is found, its `<Value>` is inserted in the tooltip.

**Comment by pastelmind** [comment]
The issue is trickier than I expected. It looks like tooltips are being rendered before `<LevelScalingArray>` is parsed.

In hero XML files, `<LevelScalingArray>` is contained in `<CUnit>`, which appears way below the tooltip elements. This poses a problem: since [`parseElements()`](../blob/1fdcad6249d78b681edb3b84ca55ff7feb874bc5/src/parser.ts#L50) parses and immediately formats each element in sequence, the contents of `<LevelScalingArray>` are parsed *after* the tooltips are generated.

Possible solutions:

1. Separate the parsing and formatting stages in `parseElements()`. Parse the entire XML file before formatting any elements. Then [`textParsers.renderTooltip()`](https://github.com/joeistas/heroes-parser/blob/1fdcad6249d78b681edb3b84ca55ff7feb874bc5/src/parsers/text-parsers.ts#L87) can be refactored into an `ElementFormatter`, which is guaranteed to run after `<LevelScalingArray>` has been parsed.
2. Somehow extract the `<LevelScalingArray>` info while rendering the tooltip. I feel this defeats the purpose of [`elementParsers.levelScalingParser()`](https://github.com/joeistas/heroes-parser/blob/1fdcad6249d78b681edb3b84ca55ff7feb874bc5/src/parsers/element-parsers.ts#L19), but this would have less side effects.

**Comment by joeistas** [comment]
The intended structure of the parsing functions are to be fairly self contained and make as few assumptions about the structure of elements as possible. heroes-parser out of the box supports parsing heroes and mounts with config options to parse any catalog of elements.

Some of the issues I see with the solution presented in PR #6 is that is assumes that the desired parent is always a Hero and it doesn't try the hero that is currently being parsed with where it is pulling the scaling info. It also assumes that one would always want to add the scaling information after the value as text. I can envision scenarios where you would want to calculate the data for each level or maybe format the text differently for each locale.

I think my preferred approach right now would be:
 - Factor out element value lookup from `renderTooltipData` into a separate function in tooltip.ts and add a new parse function for adding to element functions. The value should be added as property `value` on `TooltipReference`
 - Factor out formula calculation from `renderTooltipData` into a separate function and add a new parse function for adding to element functions. The results could be added as `formulaResults` on `TooltipData` with the key being the formula name.
 - Convert `renderTooltipData` for use as a `formatElement` function that only replaces formulas with the value in `results`.

Then in your project you can add:
 - `postParse` function for `CHero` that traverses the hero element looking for tooltip data references adds scaling information. (See `CHero` in `element-functions/base.ts` and `levelScalingParser` in `parsers/element-parsers.ts` for an example)
- `formatElement` function that is added just before `renderTooltipData` where you can format formula results before the tooltip is rendered.

For now, I would like to keep the scaling specific parts outside of this package, but I think there are changes that can be made to the tooltip parsing and rendering to make creating the feature easier.

**Comment by pastelmind** [comment]
Thank you for your feedback. I admit #6 is a dirty hack, and your suggestion makes more sense.

>Factor out element value lookup from `renderTooltipData` into a separate function in tooltip.ts and add a new parse function for adding to element functions. The value should be added as property `value` on `TooltipReference`

This would solve many problems. Currently (v1.0.11), references generated with `--profile detailed` often point to values missing in the generated JSON files. This was the whole reason I ditched writing my own formula parsing script and began digging into `renderTooltipData()`.

>`postParse` function for `CHero` that traverses the hero element looking for tooltip data references adds scaling information.

I see a potential problem with this. Blizzard uses XML formatted text all over the place--skill button tooltips, talent button tooltips, hero descriptions, etc. Is there a generic element traversal method that can search for tooltip data references?

**Comment by joeistas** [comment]
This refactor is available in v1.1
